File: sdk/javascript/src/voice.ts
/**
 * Voice/Retell client for Ragamuffin SDK
 */

import type { RagamuffinClient } from './client';
import type {
  RetellStatus,
  RetellAgent,
  WebCallResponse,
  Call,
} from './types';

export interface WebCallOptions {
  metadata?: Record<string, unknown>;
  dynamicVariables?: Record<string, string>;
}

export interface PhoneCallOptions {
  fromPhone?: string;
  metadata?: Record<string, unknown>;
}

export class VoiceClient {
  private client: RagamuffinClient;

  constructor(client: RagamuffinClient) {
    this.client = client;
  }

  /**
   * Check Retell.ai configuration status
   */
  async status(): Promise<RetellStatus> {
    return this.client.request<RetellStatus>('GET', '/retell/status');
  }

  /**
   * List all Retell agents
   */
  async agents(): Promise<{ agents: RetellAgent[] }> {
    return this.client.request<{ agents: RetellAgent[] }>('GET', '/retell/agents');
  }

  /**
   * Get a specific agent
   */
  async getAgent(agentId: string): Promise<RetellAgent> {
    return this.client.request<RetellAgent>('GET', `/retell/agents/${agentId}`);
  }

  /**
   * Create a web call
   */
  async createWebCall(
    agentId: string,
    options: WebCallOptions = {}
  ): Promise<WebCallResponse> {
    return this.client.request<WebCallResponse>('POST', '/retell/web-call', {
      body: {
        agent_id: agentId,
        metadata: options.metadata,
        dynamic_variables: options.dynamicVariables,
      },
    });
  }

  /**
   * Create a phone call
   */
  async createPhoneCall(
    agentId: string,
    toPhone: string,
    options: PhoneCallOptions = {}
  ): Promise<Call> {
    return this.client.request<Call>('POST', '/retell/phone-call', {
      body: {
        agent_id: agentId,
        to_phone: toPhone,
        from_phone: options.fromPhone,
        metadata: options.metadata,
      },
    });
  }

  /**
   * List call history
   */
  async calls(options: { limit?: number; offset?: number } = {}): Promise<{ calls: Call[] }> {
    const params = new URLSearchParams();
    if (options.limit) params.append('limit', String(options.limit));
    if (options.offset) params.append('offset', String(options.offset));
    
    const query = params.toString();
    const path = query ? `/retell/calls?${query}` : '/retell/calls';
    
    return this.client.request<{ calls: Call[] }>('GET', path);
  }

  /**
   * Get a specific call
   */
  async getCall(callId: string): Promise<Call> {
    return this.client.request<Call>('GET', `/retell/calls/${callId}`);
  }

  /**
   * End a call
   */
  async endCall(callId: string): Promise<{ message: string }> {
    return this.client.request<{ message: string }>('POST', `/retell/end-call/${callId}`);
  }

  /**
   * List available voices
   */
  async voices(): Promise<{ voices: { voice_id: string; voice_name: string }[] }> {
    return this.client.request<{ voices: { voice_id: string; voice_name: string }[] }>(
      'GET',
      '/retell/voices'
    );
  }
}

File: sdk/python/ragamuffin/voice.py
"""
Voice/Retell.ai client for Ragamuffin SDK.
"""

from typing import TYPE_CHECKING, Optional, List

if TYPE_CHECKING:
    from .client import RagamuffinClient


class VoiceClient:
    """
    Voice/Retell.ai operations for Ragamuffin API.
    
    Example:
        >>> # Check Retell status
        >>> status = client.voice.status()
        
        >>> # List agents
        >>> agents = client.voice.agents()
        
        >>> # Start a web call
        >>> call = client.voice.create_web_call("agent_id")
    """

    def __init__(self, client: "RagamuffinClient"):
        self._client = client

    def status(self) -> dict:
        """
        Check Retell.ai configuration status.
        
        Returns:
            Configuration status (configured, api_key_set, etc.)
        """
        return self._client.request("GET", "/retell/status")

    def agents(self) -> dict:
        """
        List all Retell.ai agents.
        
        Returns:
            List of agent configurations
        """
        return self._client.request("GET", "/retell/agents")

    def get_agent(self, agent_id: str) -> dict:
        """
        Get details for a specific agent.
        
        Args:
            agent_id: Agent ID
        
        Returns:
            Agent configuration and details
        """
        return self._client.request("GET", f"/retell/agents/{agent_id}")

    def create_web_call(
        self,
        agent_id: str,
        metadata: Optional[dict] = None,
        dynamic_variables: Optional[dict] = None,
    ) -> dict:
        """
        Create a browser-based voice call.
        
        Args:
            agent_id: Retell agent ID
            metadata: Optional call metadata
            dynamic_variables: Optional dynamic variables for the agent
        
        Returns:
            Call information including access_token for web SDK
        """
        data = {"agent_id": agent_id}
        if metadata:
            data["metadata"] = metadata
        if dynamic_variables:
            data["dynamic_variables"] = dynamic_variables
        
        return self._client.request("POST", "/retell/web-call", json=data)

    def create_phone_call(
        self,
        agent_id: str,
        to_phone: str,
        from_phone: Optional[str] = None,
        metadata: Optional[dict] = None,
    ) -> dict:
        """
        Create an outbound phone call.
        
        Args:
            agent_id: Retell agent ID
            to_phone: Destination phone number (E.164 format)
            from_phone: Optional source phone number
            metadata: Optional call metadata
        
        Returns:
            Call information
        """
        data = {
            "agent_id": agent_id,
            "to_phone": to_phone,
        }
        if from_phone:
            data["from_phone"] = from_phone
        if metadata:
            data["metadata"] = metadata
        
        return self._client.request("POST", "/retell/phone-call", json=data)

    def calls(
        self,
        limit: int = 50,
        offset: int = 0,
    ) -> dict:
        """
        List call history.
        
        Args:
            limit: Maximum number of calls to return
            offset: Offset for pagination
        
        Returns:
            List of calls with metadata
        """
        params = {"limit": limit, "offset": offset}
        return self._client.request("GET", "/retell/calls", params=params)

    def get_call(self, call_id: str) -> dict:
        """
        Get details for a specific call.
        
        Args:
            call_id: Call ID
        
        Returns:
            Call details including transcript if available
        """
        return self._client.request("GET", f"/retell/calls/{call_id}")

    def end_call(self, call_id: str) -> dict:
        """
        End an ongoing call.
        
        Args:
            call_id: Call ID to end
        
        Returns:
            Confirmation response
        """
        return self._client.request("POST", f"/retell/end-call/{call_id}")

    def voices(self) -> dict:
        """
        List available voices.
        
        Returns:
            List of voice options
        """
        return self._client.request("GET", "/retell/voices")
